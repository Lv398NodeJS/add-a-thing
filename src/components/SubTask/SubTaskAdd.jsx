import React, { Component } from 'react';
import { InputGroup, Button, FormControl } from 'react-bootstrap';

class SubTaskAdd extends Component {
  render() {
    const { addSubTask } = this.props;
    return (
      <InputGroup className="mb-3">
        <InputGroup.Prepend>
          <Button
            variant="outline-success"
            size="sm"
            onClick={() => {
              addSubTask(this.input.value);
              this.input.value = '';
            }}
          >
            +
          </Button>
        </InputGroup.Prepend>
        <FormControl
          size="sm"
          maxLength={40}
          ref={(input) => {
            this.input = input;
          }}
          onKeyPress={(event) => {
            if (event.key === 'Enter') {
              addSubTask(this.input.value);
              this.input.value = '';
            }
          }}
        />
      </InputGroup>
    );
  }
}

export default SubTaskAdd;
