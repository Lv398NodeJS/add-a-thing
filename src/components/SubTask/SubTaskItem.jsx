import React, { Component } from 'react';
import {
  Button, Form, Row, Col,
} from 'react-bootstrap';
import './index.scss';

export default class Subtask extends Component {
  render() {
    const {
      subTask: { text, completed, key },
      changeSubTaskStatus,
      deleteSubTask,
    } = this.props;
    return (
      <Row className="mb-1">
        <Col className="d-flex justify-content-sm-start col-sm-11">
          <Form.Check type="checkbox" custom id={key}>
            <Form.Check.Input
              type="checkbox"
              checked={completed}
              onChange={() => changeSubTaskStatus(key)}
            />
            <Form.Check.Label className={completed && 'task-completed'}>{text}</Form.Check.Label>
          </Form.Check>
        </Col>
        <Col className="d-flex justify-content-sm-end col-sm-1">
          <Button
            variant="outline-danger"
            size="sm"
            as="input"
            value="x"
            onClick={() => deleteSubTask(key)}
          />
        </Col>
      </Row>
    );
  }
}
