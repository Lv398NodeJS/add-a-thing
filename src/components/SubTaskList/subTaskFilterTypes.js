export const subtaskFilterTypes = {
  SHOW_ALL: 'SHOW_ALL',
  SHOW_COMPLETED: 'SHOW_COMPLETED',
  SHOW_ACTIVE: 'SHOW_ACTIVE',
};
