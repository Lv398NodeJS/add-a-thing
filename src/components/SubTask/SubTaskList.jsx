/* eslint-disable no-param-reassign */
import React, { Component } from 'react';
import { Container, Row, Col } from 'react-bootstrap';
import './index.scss';
import uuidv1 from 'uuid/v1';
import SubTaskItem from './SubTaskItem';
import SubTaskAdd from './SubTaskAdd';
import SubTaskProgressBar from './SubTaskProgressBar';

class SubTaskList extends Component {
  constructor(props) {
    super(props);

    const { subtasks } = this.props;
    this.state = { subTasks: subtasks };

    this.changeSubTaskStatus = this.changeSubTaskStatus.bind(this);
    this.addSubTask = this.addSubTask.bind(this);
    this.deleteSubTask = this.deleteSubTask.bind(this);
  }

  changeSubTaskStatus(id) {
    this.setState((prevState) => {
      const updatedSubTasks = prevState.subTasks.map((subTask) => {
        if (subTask.id === id) subTask.completed = !subTask.completed;
        return subTask;
      });
      return { subTasks: updatedSubTasks };
    });
  }

  addSubTask(subTaskText) {
    subTaskText = subTaskText.trim();
    this.setState(prevState => ({
      subTasks: [...prevState.subTasks, { id: uuidv1(), completed: false, text: subTaskText }],
    }));
  }

  deleteSubTask(subTaskId) {
    this.setState((prevState) => {
      const updatedSubTasks = prevState.subTasks.filter(subTask => subTask.id !== subTaskId);
      return { subTasks: updatedSubTasks };
    });
  }

  render() {
    const { subTasks } = this.state;
    const subTaskItems = subTasks.map(subTask => (
      <SubTaskItem
        key={subTask.id}
        subTask={subTask}
        changeSubTaskStatus={this.changeSubTaskStatus}
        deleteSubTask={this.deleteSubTask}
      />
    ));

    return (
      <Container>
        <SubTaskProgressBar subTasks={subTasks} />
        {subTaskItems}
        <Row className="justify-content-sm-center">
          <Col>
            <SubTaskAdd addSubTask={this.addSubTask} />
          </Col>
        </Row>
      </Container>
    );
  }
}

export default SubTaskList;
