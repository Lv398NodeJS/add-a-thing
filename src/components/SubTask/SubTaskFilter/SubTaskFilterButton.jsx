import React, { Component } from 'react';
import { Button } from 'react-bootstrap';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as subTaskActions from '../../../actions/subTaskActions';
import { subtaskFilterTypes } from '../subTaskFilterTypes';

class SubTaskFilterButton extends Component {
  getTitle = () => {
    const { type } = this.props;
    switch (type) {
      case subtaskFilterTypes.SHOW_ACTIVE:
        return 'Active';
      case subtaskFilterTypes.SHOW_COMPLETED:
        return 'Completed';
      default:
        return 'All';
    }
  };

  isActive = () => {
    const { type, currentFilter } = this.props;
    if (type === currentFilter) return true;
    return false;
  }

  render() {
    const { type, subTaskActions: { setSubTaskFilter } } = this.props;
    return (
      <Button size="sm" onClick={() => setSubTaskFilter(type)} active={this.isActive()}>{this.getTitle()}</Button>
    );
  }
}
const mapStateToProps = ({ subTaskListReducer: { currentFilter } }) => ({
  currentFilter,
});
const mapDispatchToProps = dispatch => ({
  subTaskActions: bindActionCreators(subTaskActions, dispatch),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(SubTaskFilterButton);
