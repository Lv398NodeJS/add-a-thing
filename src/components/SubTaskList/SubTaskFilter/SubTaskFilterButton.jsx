import React, { Component } from 'react';
import { Button } from 'react-bootstrap';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as subTaskListActions from '../../../actions/subTaskListActions';
import { subtaskFilterTypes } from '../subTaskFilterTypes';

class SubTaskFilterButton extends Component {
  getTitle = () => {
    const { filter } = this.props;
    switch (filter) {
      case subtaskFilterTypes.SHOW_ACTIVE:
        return 'Active';
      case subtaskFilterTypes.SHOW_COMPLETED:
        return 'Completed';
      default:
        return 'All';
    }
  };

  render() {
    const { filter, subTaskListActions: { setSubTaskFilter } } = this.props;
    return (
      <Button size="sm" onClick={() => setSubTaskFilter(filter)}>{this.getTitle()}</Button>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  subTaskListActions: bindActionCreators(subTaskListActions, dispatch),
});

export default connect(
  null,
  mapDispatchToProps,
)(SubTaskFilterButton);
