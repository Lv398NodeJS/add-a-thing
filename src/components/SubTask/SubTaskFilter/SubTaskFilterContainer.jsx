import React, { Component } from 'react';
import { ButtonGroup } from 'react-bootstrap';
import { SHOW_ALL, SHOW_ACTIVE, SHOW_COMPLETED } from '../subTaskFilterTypes';
import SubTaskFilterButton from './SubTaskFilterButton';

class SubTaskFilterContainer extends Component {
  render() {
    return (
      <>
        <span>Show: </span>
        <ButtonGroup>
          <SubTaskFilterButton filterType={SHOW_ALL} />
          <SubTaskFilterButton filterType={SHOW_ACTIVE} />
          <SubTaskFilterButton filterType={SHOW_COMPLETED} />
        </ButtonGroup>
      </>
    );
  }
}


export default SubTaskFilterContainer;
