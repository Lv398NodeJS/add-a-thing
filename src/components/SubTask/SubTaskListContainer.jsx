import React, { Component } from 'react';
import { Row, Col, Container } from 'react-bootstrap';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as subTaskActions from '../../actions/subTaskListActions';
import SubTaskProgressBar from './SubTaskProgressBar/SubTaskProgressBar';
import SubTaskList from './SubTaskList/SubTaskList';
import SubTaskFilterContainer from './SubTaskFilter/SubTaskFilterContainer';
import SubTaskAdd from './SubTaskAdd/SubTaskAdd';

export class SubTaskListContainer extends Component {
  render() {
    const { taskRef, taskStatus, subtaskList = {} } = this.props;
    return (
      <Container fluid className="my-0 mx-0">
        <Row>
          <Col className="col-4">
            <h5>Subtasks:</h5>
          </Col>
          <Col className="col-8 text-right">
            <SubTaskFilterContainer />
          </Col>
        </Row>
        <SubTaskProgressBar subtaskList={subtaskList} />
        <SubTaskList taskRef={taskRef} />
        <SubTaskAdd taskStatus={taskStatus} taskRef={taskRef} />
      </Container>
    );
  }
}

const mapStateToProps = ({ subTaskListReducer: { taskStatus, subtaskList } }) => ({
  taskStatus,
  subtaskList,
});

const mapDispatchToProps = dispatch => ({
  subTaskListActions: bindActionCreators(subTaskActions, dispatch),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(SubTaskListContainer);
