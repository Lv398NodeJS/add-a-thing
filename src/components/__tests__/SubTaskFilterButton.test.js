import React from 'react';
import { shallow } from 'enzyme';
import { SubTaskFilterButtonComponent } from '../SubTask/SubTaskFilter/SubTaskFilterButton';
import { subtaskFilterTypes } from '../SubTask/subTaskFilterTypes';

describe('SubTaskFilterButton component', () => {
  const subTaskActions = { setSubTaskFilter: jest.fn() };
  it('should render correctly and match the snapshot', () => {
    const subTaskFilterButtonComponent = shallow(
      <SubTaskFilterButtonComponent subTaskActions={subTaskActions} />,
    );
    expect(subTaskFilterButtonComponent).toMatchSnapshot();
  });

  it('should render Bootstrap 4 Button', () => {
    const subTaskFilterButtonComponent = shallow(
      <SubTaskFilterButtonComponent subTaskActions={subTaskActions} />,
    );
    expect(subTaskFilterButtonComponent.find('Button').length).toBe(1);
  });

  it('should render Bootstrap 4 Button with text depending on "filterType" prop', () => {
    const subTaskFilterButtonComponent = shallow(
      <SubTaskFilterButtonComponent
        subTaskActions={subTaskActions}
        filterType={subtaskFilterTypes.SHOW_ALL}
      />,
    );
    expect(subTaskFilterButtonComponent.find('Button').text()).toBe('All');
  });

  it('should render Bootstrap 4 Button with text depending on "filterType" prop', () => {
    const subTaskFilterButtonComponent = shallow(
      <SubTaskFilterButtonComponent
        subTaskActions={subTaskActions}
        filterType={subtaskFilterTypes.SHOW_ACTIVE}
      />,
    );
    expect(subTaskFilterButtonComponent.find('Button').text()).toBe('Active');
  });

  it('should render Bootstrap 4 Button with text depending on "filterType" prop', () => {
    const subTaskFilterButtonComponent = shallow(
      <SubTaskFilterButtonComponent
        subTaskActions={subTaskActions}
        filterType={subtaskFilterTypes.SHOW_COMPLETED}
      />,
    );
    expect(subTaskFilterButtonComponent.find('Button').text()).toBe('Completed');
  });

  it('should render active Bootstrap 4 Button if prop "filterType" is equal to prop "currentFilter"', () => {
    const subTaskFilterButtonComponent = shallow(
      <SubTaskFilterButtonComponent
        subTaskActions={subTaskActions}
        filterType={subtaskFilterTypes.SHOW_COMPLETED}
        currentFilter={subtaskFilterTypes.SHOW_COMPLETED}
      />,
    );
    expect(subTaskFilterButtonComponent.find('Button').prop('active')).toBeTruthy();
  });
});
