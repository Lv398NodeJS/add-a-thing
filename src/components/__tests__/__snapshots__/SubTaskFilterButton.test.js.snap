// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubTaskFilterButton component should render correctly and match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SubTaskFilterButton
    subTaskActions={
      Object {
        "setSubTaskFilter": [MockFunction],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "active": true,
      "children": "All",
      "disabled": false,
      "onClick": [Function],
      "size": "sm",
      "type": "button",
      "variant": "primary",
    },
    "ref": null,
    "rendered": "All",
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "defaultProps": Object {
        "active": false,
        "disabled": false,
        "type": "button",
        "variant": "primary",
      },
      "displayName": "Button",
      "render": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "active": true,
        "children": "All",
        "disabled": false,
        "onClick": [Function],
        "size": "sm",
        "type": "button",
        "variant": "primary",
      },
      "ref": null,
      "rendered": "All",
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "defaultProps": Object {
          "active": false,
          "disabled": false,
          "type": "button",
          "variant": "primary",
        },
        "displayName": "Button",
        "render": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
